// @faustwp/core/dist/mjs/getProps.js
import isBoolean from 'lodash/isBoolean.js';
import isObject from 'lodash/isObject.js';
import { addApolloState, getApolloClient } from './client.js';
/**
 * This helper function lets you build a static site with your WordPress data
 *
 * @param {GetStaticPropsContext} context
 * @param {GetNextStaticPropsConfig} cfg
 */
export async function getNextStaticProps(context, cfg) {
    const { notFound, redirect, Page, props } = cfg;
    const apolloClient = getApolloClient();
    if (isBoolean(notFound) && notFound === true) {
        return {
            notFound,
        };
    }
    if (isObject(redirect)) {
        return {
            redirect,
        };
    }
    if (Page.query) {
        await apolloClient.query({
            query: Page.query,
            variables: (Page === null || Page === void 0 ? void 0 : Page.variables) ? Page === null || Page === void 0 ? void 0 : Page.variables(context) : undefined,
        });
    }
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return addApolloState(apolloClient, { props: Object.assign({}, props) });
}
/**
 * This helper function lets you server side render your page with WordPress data
 *
 * @param {GetServerSidePropsContext} context
 * @param {GetNextServerSidePropsConfig} cfg
 */
export async function getNextServerSideProps(context, cfg) {
    const { res } = context;
    const { notFound, redirect, Page, props } = cfg;
    const apolloClient = getApolloClient();
    res.setHeader('x-powered-by', 'Faust');
    if (isBoolean(notFound) && notFound === true) {
        return {
            notFound,
        };
    }
    if (isObject(redirect)) {
        return {
            redirect,
        };
    }
    if (Page.query) {
        await apolloClient.query({
            query: Page.query,
            variables: (Page === null || Page === void 0 ? void 0 : Page.variables) ? Page === null || Page === void 0 ? void 0 : Page.variables(context) : undefined,
        });
    }
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return addApolloState(apolloClient, {
        props: Object.assign({}, props),
    });
}
