// @faustwp/core/dist/mjs/getWordPressProps.js
import { SEED_QUERY } from './queries/seedQuery.js';
import { getTemplate } from './getTemplate.js';
import { addApolloState, getApolloClient } from './client.js';
import { getConfig } from './config/index.js';
import { hooks } from './hooks/index.js';
function isSSR(ctx) {
    return ctx.req !== undefined;
}
export async function getWordPressProps(options) {
    var _a, _b, _c;
    const { templates } = getConfig();
    if (!templates) {
        throw new Error('Templates are required. Please add them to your config.');
    }
    const { ctx } = options;
    const client = getApolloClient();
    let resolvedUrl = null;
    if (!isSSR(ctx)) {
        const wordPressNodeParams = (_a = ctx.params) === null || _a === void 0 ? void 0 : _a.wordpressNode;
        if (wordPressNodeParams && Array.isArray(wordPressNodeParams)) {
            resolvedUrl = `/${wordPressNodeParams.join('/')}`;
        }
        else {
            resolvedUrl = '/';
        }
    }
    else {
        resolvedUrl = ctx.req.url;
        ctx.res.setHeader('x-powered-by', 'Faust');
    }
    if (!resolvedUrl) {
        return {
            notFound: true,
        };
    }
    const seedQuery = hooks.applyFilters('seedQueryDocumentNode', SEED_QUERY, {
        resolvedUrl,
    });
    const seedQueryRes = await client.query({
        query: seedQuery,
        variables: { uri: resolvedUrl },
    });
    const seedNode = (_b = seedQueryRes === null || seedQueryRes === void 0 ? void 0 : seedQueryRes.data) === null || _b === void 0 ? void 0 : _b.node;
    if (!seedNode) {
        return {
            notFound: true,
        };
    }
    const template = getTemplate(seedNode, templates);
    if (!template) {
        return {
            notFound: true,
        };
    }
    let templateQueryRes;
    if (template.query) {
        templateQueryRes = await client.query({
            query: template.query,
            variables: (template === null || template === void 0 ? void 0 : template.variables)
                ? template === null || template === void 0 ? void 0 : template.variables(seedNode, { asPreview: false })
                : undefined,
        });
    }
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return addApolloState(client, {
        props: {
            /**
             * The following props may be null coalesced as an "undefined"
             * value is not able to be serialized
             */
            __SEED_NODE__: seedNode !== null && seedNode !== void 0 ? seedNode : null,
            __TEMPLATE_QUERY_DATA__: (_c = templateQueryRes === null || templateQueryRes === void 0 ? void 0 : templateQueryRes.data) !== null && _c !== void 0 ? _c : null,
        },
    });
}
